File: app/src/main/java/com/sample/app/repository/CustomerRepository.java
package com.sample.app.repository;

import android.util.Log;

import com.sample.app.BaseApplication;
import com.sample.app.db.greendao.CustomerDao;
import com.sample.app.entity.Customer;

import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

public class CustomerRepository {
    private static AtomicLong sId = new AtomicLong();

    private static CustomerDao getDao() {
        return BaseApplication.getDaoSession().getCustomerDao();
    }

    public static void saveAll(List<Customer> itemList) {
        CustomerDao dao = getDao();

        List<Customer> oldItemList;
        Customer old;
        for (Customer item : itemList) {
            oldItemList = dao.queryBuilder()
                    .where(CustomerDao.Properties.Name.eq(item.getName())).list();
            if (!oldItemList.isEmpty()) {
                old = oldItemList.get(0);
                item.setId(old.getId());
                dao.update(item);
                sId.incrementAndGet();
            } else {
                item.setId(sId.getAndIncrement());
                insert(item, dao);
            }
        }
    }

    private static void insert(Customer item, CustomerDao dao) {
        try {
            dao.insert(item);
        } catch (Throwable throwable) {
            Log.e("dao.insert(Customer)", throwable.toString());
            item.setId(sId.incrementAndGet());
            insert(item, dao);
        }
    }

    public static List<Customer> search(String searchStr) {
        return getDao().queryBuilder().where(
                CustomerDao.Properties.Name.like("%" + searchStr + "%")).list();
    }

    public static List<Customer> getAll() {
        return getDao().loadAll();
    }
}

File: app/src/main/java/com/sample/app/repository/CustomerRepository.kt
package com.sample.app.repository

import android.util.Log

import com.sample.app.BaseApplication
import com.sample.app.db.greendao.CustomerDao
import com.sample.app.entity.Customer
import java.util.concurrent.atomic.AtomicLong

object CustomerRepository {
    private val sId = AtomicLong()

    private val dao: CustomerDao
        get() = BaseApplication.daoSession!!.customerDao

    fun saveAll(itemList: List<Customer>) {
        val dao = dao

        var oldItemList: List<Customer>
        var old: Customer
        for (item in itemList) {
            oldItemList = dao.queryBuilder()
                    .where(CustomerDao.Properties.Name.eq(item.name)).list()
            if (!oldItemList.isEmpty()) {
                old = oldItemList[0]
                item.id = old.id
                dao.update(item)
                sId.incrementAndGet()
            } else {
                item.id = sId.getAndIncrement()
                insert(item, dao)
            }
        }
    }

    private fun insert(item: Customer, dao: CustomerDao) {
        try {
            dao.insert(item)
        } catch (throwable: Throwable) {
            Log.e("dao.insert(Customer)", throwable.toString())
            item.id = sId.incrementAndGet()
            insert(item, dao)
        }

    }

    fun search(searchStr: String): List<Customer> {
        return dao.queryBuilder().where(
                CustomerDao.Properties.Name.like("%$searchStr%")).list()
    }

    val all: List<Customer>
        get() = dao.loadAll()
}

